### PR TITLE
Make Array Send by using `Arc<Mutext<RawContext>>`

### PR DESCRIPTION
**Pull request by davisp**
This makes Array Send by putting the RawContext into an `Arc<Mutext<T>>`.

### CODE REVIEW

File: tiledb/api/src/context.rs [hunk 1]
```diff
@@ -68,95 +68,110 @@ impl Context {
...
-            )
+    pub fn capi_call<Callable>(&self, action: Callable) -> TileDBResult<()>
+    where
+        Callable: FnOnce(*mut ffi::tiledb_ctx_t) -> i32,
+    {
+        // Scope the lock so we can get the error if one happens.
+        let res = {
+            // Docs say that the error result from a poisoned lock are usually
+            // just propagated to all threads by just calling unwrap on locks.
+            let raw_ctx = self.raw.lock().unwrap();
```
Comment (rroelke) [inline tiledb/api/src/context.rs:86]: Per my above comment I think this should be reverted, but...

It is also done incorrectly as the lock must be held before `action` begins and released after `expect_last_error` ends.

File: tiledb/api/src/context.rs [hunk 2]
```diff
@@ -68,95 +68,110 @@ impl Context {
...
+            ffi::tiledb_ctx_get_config(ctx, &mut c_cfg)
+        })?;
+
+        Ok(Config {
+            raw: RawConfig::Owned(c_cfg),
+        })
     }
 
-    pub fn get_last_error(&self) -> Option<Error> {
+    // This is a bit awkard. We *expect* this to return an Err variant of the
```
Comment (rroelke) [inline tiledb/api/src/context.rs:123]: Are your concerns allayed if these are changed to `pub(crate)`?

All the API calls should check these, there should be no reason for a user to have to do so.

File: tiledb/api/src/context.rs [hunk 3]
```diff
@@ -68,95 +68,110 @@ impl Context {
...
-            false
-        }
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_ctx_is_supported_fs(ctx, fs as u32, &mut supported)
+        })?;
+
+        Ok(supported == 1)
     }
 
     pub fn set_tag(&self, key: &str, val: &str) -> TileDBResult<()> {
```
Comment (rroelke) [inline tiledb/api/src/context.rs:161]: Based on the discussion in `#core` the other day, this should be required as part of the context construction? Should we file a story for that?

File: tiledb/api/src/context.rs (hunk 1 above)
Comment (davisp) [reply tiledb/api/src/context.rs:86]: Ohhhhh, good catch!

File: tiledb/api/src/context.rs (hunk 2 above)
Comment (davisp) [reply tiledb/api/src/context.rs:123]: I was mostly just not liking my weird use of TileDBResult here. It was just "Wait, should an error be returned as an `Ok` or an `Err`?" sort of weirdness. As you pointed out we'd want to make these called while the lock is held if we go that direction.

File: tiledb/api/src/context.rs (hunk 3 above)
Comment (davisp) [reply tiledb/api/src/context.rs:161]: Done: https://app.shortcut.com/tiledb-inc/story/46092

File: tiledb/api/src/context.rs [hunk 4]
```diff
@@ -37,27 +32,32 @@ pub trait ContextBound<'ctx> {
...
+    where
+        Callable: FnOnce(*mut ffi::tiledb_ctx_t) -> i32,
+    {
+        self.context().capi_call(action)
     }
 }
 
 pub struct Context {
-    raw: RawContext,
+    raw: Arc<Mutex<RawContext>>,
```
[the remark this answers is not in the document]
Comment (davisp) [reply tiledb/api/src/context.rs:57]: capi_call is now its own PR: #76

File: tiledb/api/src/context.rs (hunk 1 above)
Comment (davisp) [reply tiledb/api/src/context.rs]: Fixed.

File: tiledb/api/src/context.rs (hunk 2 above)
Comment (davisp) [reply tiledb/api/src/context.rs]: I did this differently and just swallowed the error like we were before so its same as it used to be.

**Comment by davisp** [comment]
Closing this as we'll not likely need it.

**Pull request closed without merging**